### ISSUE TITLE
Orthogonal initialization gives inaccurate result when running on GPU

### ISSUE THREAD

**Issue by dargilboa** [tensorflow/tensorflow#22226]
### System information
- **Have I written custom code**: see example below
- **OS Platform and Distribution**:Google Colabratory (also in compiled code on Linux)
- **Mobile device**: N/A
- **TensorFlow installed from (source or binary)**: Google Colabratory
- **TensorFlow version (use command below)**: 1.10.1
- **Python version**: 2.7.14 
- **Bazel version (if compiling from source)**:  N/A
- **GCC/Compiler version (if compiling from source)**:  N/A
- **CUDA/cuDNN version**:  N/A
- **GPU model and memory**: Issue occurred both on Telsa V100 and Tesla K80 with at least 11GB of available memory (but not when using CPU runtime)
- **Exact command to reproduce**: see example below

### Describe the problem
When generating an orthogonal matrix and running on a GPU, if the matrix is 1000x1000 or larger the resulting matrix is not orthogonal. The deviations are several orders of magnitude larger than expected from finite precision errors.
The issue doesn't occur for smaller n (say 100) or when running on a CPU.

In the example below, the final result has entry-wise deviations from the identity matrix of order 1e-2.

### Source code / logs
 

```
import tensorflow as tf
tf.enable_eager_execution()
tf.set_random_seed(0)
init = tf.orthogonal_initializer(1,seed=0)
a = init([1000,1000])
tf.matmul(a,tf.transpose(a))
```

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by dargilboa** [comment]
sure, I updated the post

**Comment by harshini-gadige** [comment]
@rohan100jain Hi, could you please post the output which you get when running on a GPU for a 1000x1000 matrix.

**Comment by dargilboa** [comment]
This is for a 1000x1000
<tf.Tensor: id=24, shape=(1000, 1000), dtype=float32, numpy=
array([[ 1.0023266 ,  0.02684135, -0.01610221, ..., -0.00190309,
         0.01394506, -0.00587076],
       [ 0.02684135,  1.0824999 , -0.0193597 , ...,  0.0389607 ,
        -0.00889984, -0.05881837],
       [-0.01610221, -0.0193597 ,  1.0005746 , ..., -0.02111715,
        -0.00168062,  0.01316965],
       ...,
       [-0.00190309,  0.0389607 , -0.02111715, ...,  1.0288322 ,
        -0.01475003, -0.02641175],
       [ 0.01394506, -0.00889984, -0.00168062, ..., -0.01475003,
         1.0051492 ,  0.01033276],
       [-0.00587076, -0.05881837,  0.01316965, ..., -0.02641175,
         0.01033276,  1.0294992 ]], dtype=float32)>

while on a CPU I get 

<tf.Tensor: id=24, shape=(1000, 1000), dtype=float32, numpy=
array([[ 9.9999940e-01, -1.8626451e-09, -2.7939677e-08, ...,
         1.5075784e-08, -8.3819032e-09,  3.7252903e-09],
       [-1.8626451e-09,  1.0000001e+00, -4.7730282e-08, ...,
         3.1664968e-08, -6.5192580e-08,  9.3132257e-10],
       [-2.7939677e-08, -4.7730282e-08,  1.0000000e+00, ...,
        -5.0291419e-08, -5.8207661e-09,  9.3132257e-09],
       ...,
       [ 1.5075784e-08,  3.1664968e-08, -5.0291419e-08, ...,
         1.0000000e+00, -3.0733645e-08, -1.8742867e-08],
       [-8.3819032e-09, -6.5192580e-08, -5.8207661e-09, ...,
        -3.0733645e-08,  9.9999988e-01, -2.7939677e-08],
       [ 3.7252903e-09,  9.3132257e-10,  9.3132257e-09, ...,
        -1.8742867e-08, -2.7939677e-08,  9.9999994e-01]], dtype=float32)>

also when comparing the matrix a in both cases, although most entries are identical between the CPU and GPU (up to finite precision errors) there are larger deviations in a few rows

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by KyotoSunshine** [comment]
Any update on this?

**Comment by mohantym** [comment]
Hi @dargilboa ! Have you checked this issue using [tf.keras.initializers.Orthogonal](https://www.tensorflow.org/api_docs/python/tf/keras/initializers/Orthogonal) yet ? I am getting same result for CPU and GPU through it. Attaching [gist](https://colab.sandbox.google.com/gist/mohantym/86d35b123facf588ac30288267eb3258/git_22226.ipynb#scrollTo=q3zITfhJFaUM) for reference.
